File: 3348-minimum-cost-walk-in-weighted-graph/minimum-cost-walk-in-weighted-graph.java
class Solution {
    int[] root;
    int[] treeHeight;
    int[] minPathCostAND;

    public int find(int node) {
        if (root[node] == node) {
            return node;
        }

        return root[node] = find(root[node]);
    }

    public void union(int rootU, int rootV) {
        if (treeHeight[rootU] > treeHeight[rootV]) {
            minPathCostAND[rootU] &= minPathCostAND[rootV];
            root[rootV] = rootU;
        } else if (treeHeight[rootU] < treeHeight[rootV]) {
            minPathCostAND[rootV] &= minPathCostAND[rootU];
            root[rootU] = rootV;
        } else {
            minPathCostAND[rootU] &= minPathCostAND[rootV];
            root[rootV] = rootU;
            treeHeight[rootU]++;
        }
    }

    public int[] minimumCost(int n, int[][] edges, int[][] query) {
        root = new int[n];
        treeHeight = new int[n];
        minPathCostAND = new int[n];

        for (int i = 0; i < n; i++) {
            treeHeight[i] = 1;
            root[i] = i;
            minPathCostAND[i] = -1;
        }

        for (int i = 0; i < edges.length; i++) {
            int nodeU = edges[i][0];
            int nodeV = edges[i][1];
            int weight = edges[i][2];

            int rootU = find(nodeU);
            int rootV = find(nodeV);

            if (rootU != rootV) {
                union(rootU, rootV);
                rootU = find(nodeU);
                rootV = find(nodeV);
            }

            minPathCostAND[rootU] &= weight;
        }

        int[] result = new int[query.length];

        for (int i = 0; i < query.length; i++) {
            int nodeU = query[i][0];
            int nodeV = query[i][1];

            int rootU = find(nodeU);
            int rootV = find(nodeV);

            if (nodeU == nodeV) {
                result[i] = 0;
            } else if (rootU != rootV) {
                result[i] = -1;
            } else {
                result[i] = minPathCostAND[rootU];
            }
        }

        return result;
    }
}

File: 3348-minimum-cost-walk-in-weighted-graph/minimum-cost-walk-in-weighted-graph.go
func find(parent []int, node int) int {
	if parent[node] == node {
		return node;
	}

	parent[node] = find(parent, parent[node]);

	return parent[node];
}

func union(parent, rank, minPathAND []int, u, v int) {
	if rank[u] > rank[v] {
		parent[v] = u;
		minPathAND[u] &= minPathAND[v];
	} else if rank[u] < rank[v] {
		parent[u] = v;
		minPathAND[v] &= minPathAND[u];
	} else {
		parent[v] = u;
		minPathAND[u] &= minPathAND[v];
		rank[u]++;
	}
}

func minimumCost(n int, edges [][]int, query [][]int) []int {
    parent := make([]int, n);
	rank := make([]int, n);
	minPathAND := make([]int, n);

	for i := 0; i < n; i++ {
		parent[i] = i;
		rank[i] = 1;
		minPathAND[i] = -1;
	}

	for _, edge := range edges {
		u := edge[0];
		v := edge[1];
		weight := edge[2];

		rootU := find(parent, u);
		rootV := find(parent, v);

		if rootU != rootV {
			union(parent, rank, minPathAND, rootU, rootV);
			rootU = find(parent, u);
		}

		minPathAND[rootU] &= weight;
	}

	result := make([]int, len(query));

	for i, q := range query {
		u := q[0];
		v := q[1];

		rootU := find(parent, u);
		rootV := find(parent, v);

		if u == v {
			result[i] = 0;
		} else if rootU != rootV {
			result[i] = -1;
		} else {
			result[i] = minPathAND[rootU];
		}
	}

	return result;
}